### PR TITLE
Implement critical hits

### PR DESCRIPTION
**Pull request by AndreasHGK**
This has been tested

### CODE REVIEW

File: server/player/player.go [hunk 1]
```diff
@@ -1310,8 +1310,18 @@ func (p *Player) AttackEntity(e world.Entity) {
 
 	force, height := 0.45, 0.3608
 
+	critical := false
+	insideBlock := p.World().Block(cube.PosFromVec3(p.Position()))
+	_, isLadder := insideBlock.(block.Ladder)
+	_, isWater := insideBlock.(block.Water)
```
Comment (T14Raptor) [inline server/player/player.go:1316]: These block checks are not needed due to the fall distance check.

File: server/player/handler.go [hunk 2]
```diff
@@ -83,11 +83,12 @@ type Handler interface {
 	// may be called to cancel the attack, which will cancel damage dealt to the target and will stop the
 	// entity from being knocked back.
 	// The entity attacked may not be alive (implements entity.Living), in which case no damage will be dealt
-	// and the target won't be knocked back.
+	// and the target won't be knocked back, nor will the critical hit particles be displayed.
 	// The entity attacked may also be immune when this method is called, in which case no damage and knock-
 	// back will be dealt.
 	// The knock back force and height is also provided which can be modified.
-	HandleAttackEntity(ctx *event.Context, e world.Entity, force, height *float64)
+	// The attack can be a critical attack
```
Comment (T14Raptor) [inline server/player/handler.go:90]: Could you give more info on what a critical attack does here?

File: server/player/player.go [hunk 3]
```diff
@@ -1310,8 +1310,14 @@ func (p *Player) AttackEntity(e world.Entity) {
 
 	force, height := 0.45, 0.3608
 
+	critical := false
+	// todo: check if player has slow falling or blindness when we can
+	if !p.Flying() && !p.OnGround() && p.FallDistance() > 0 {
```
Comment (T14Raptor) [inline server/player/player.go:1315]: This if statement can be inlined, like `critical := !p.Flying() …`

File: server/session/world.go [hunk 4]
```diff
@@ -625,6 +625,13 @@ func (s *Session) ViewEntityAction(e world.Entity, a action.Action) {
 			EntityRuntimeID: s.entityRuntimeID(e),
 			EventType:       packet.ActorEventHurt,
 		})
+	case action.CriticalHit:
+		s.writePacket(
```
Comment (T14Raptor) [inline server/session/world.go:629]: Fix inconsistent formatting here

File: server/player/handler.go (hunk 2 above)
Comment (AndreasHGK) [reply server/player/handler.go]: Woops, i forgot to do this 😅

File: server/player/player.go [hunk 5]
```diff
@@ -1310,8 +1310,11 @@ func (p *Player) AttackEntity(e world.Entity) {
 
 	force, height := 0.45, 0.3608
 
+	// todo: check if player has slow falling or blindness when we can
```
Comment (T14Raptor) [inline server/player/player.go:1313]: This can be done now

File: server/player/player.go [hunk 6]
```diff
@@ -1310,8 +1310,18 @@ func (p *Player) AttackEntity(e world.Entity) {
 
 	force, height := 0.45, 0.3608
 
+	critical := !p.Flying() && !p.OnGround() && p.FallDistance() > 0
+	sf := effect.SlowFalling{}
+	bl := effect.Blindness{}
```
Comment (T14Raptor) [inline server/player/player.go:1315]: I would just inline these

**Review by Sandertv — APPROVED** [review]
This looks all good to me, has it been tested?

**Comment by AndreasHGK** [comment]
it has been tested yes

**Pull request merged**